### PR TITLE
Add support for DECSCUSR "0" to restore cursor to user default 

### LINKED ISSUE #1604: Consider supporting [extension to VT,] DECSCUSR "0" to restore cursor to user default to help vim/others
# Environment

```none
Windows build number: 1903
Windows Terminal version (if applicable): 0.2.1715

Any other software? neovim v0.3.4
```

# Steps to reproduce
- Create a profile that uses cursorShape "vintage"
- Open nvim
- close nvim

# Expected behavior
After closing nvim I would expect to have the "vintage" cursor again.

# Actual behavior
Cursor is a filledBox that doesn't blink anymore

### PR DESCRIPTION
**Pull request by skyline75489**
This PR is about the behavior of DECSCUSR. This PR changes the meaning
of DECSCUSR 0 to restore the cursor style back to user default. This
differs from what VT spec says but it’s used in popular terminal
emulators like iTerm2 and VTE-based ones. See #1604. 

Another change is that for parameter greater than 6, DECSCUSR should be
ignored, instead of restoring the cursor to legacy. This PR fixes it.
See #7382.

Fixes #1604.

### CODE REVIEW

**Comment by skyline75489** [comment]
I don't think this actually *fixes* #1604 as vim does not attempt to restore cursor shape on exit. So..what do you guys think?

**Comment by skyline75489** [comment]
Hey seems we got VS 16.7 for CI build now?

File: src/terminal/adapter/DispatchTypes.hpp
Comment (j4james) [inline src/terminal/adapter/DispatchTypes.hpp:220]: Instead of adding a comment here, couldn't we just rename the enum value to something like `UserDefault`, and then also rename `BlinkingBlockDefault` to `BlinkingBlock`. That seems like it would be a better indication of what these values actually represent.

File: src/cascadia/TerminalCore/TerminalApi.cpp
Comment (j4james) [inline src/cascadia/TerminalCore/TerminalApi.cpp:436]: Why do we need a new method for this? Wouldn't it be easier to just add it to the switch statement in the existing `SetCursorStyle` method? That also raises the question of why this is resetting values that the `SetCursorStyle` method never changed. I would have thought the only thing it should be changing is the type and blink status?

File: src/terminal/adapter/adaptDispatch.cpp
Comment (j4james) [inline src/terminal/adapter/adaptDispatch.cpp:2232]: Again I don't think we need a separate method for this, and ideally conhost would support some version of the "restore default" functionality too, but worst case the existing behaviour should remain unchanged. We definitely shouldn't be returning false without doing anything.

File: src/terminal/parser/OutputStateMachineEngine.cpp
Comment (j4james) [inline src/terminal/parser/OutputStateMachineEngine.cpp:675]: None of this would be needed if you just let `SetCursorStyle` handle the option.

File: src/terminal/parser/OutputStateMachineEngine.hpp
Comment (j4james) [inline src/terminal/parser/OutputStateMachineEngine.hpp:241]: If you go with my suggested renaming of the enums, don't forget to change this to `UserDefault`, or whatever you decide for the name.

**Review by j4james — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalCore/TerminalApi.cpp
Comment (skyline75489) [inline src/cascadia/TerminalCore/TerminalApi.cpp:393]: Is there a way to know whether blinking should be allowed, without `switch` on `_defaultCursorShape`?

File: src/terminal/adapter/DispatchTypes.hpp
Comment (skyline75489) [inline src/terminal/adapter/DispatchTypes.hpp:219]: Thanks @j4james . You're right. This is much better.

File: src/cascadia/TerminalCore/TerminalApi.cpp
Comment (j4james) [reply src/cascadia/TerminalCore/TerminalApi.cpp:393]: In Windows Terminal I think the default for _shouldBlink_ would always be true (at least for now), since there is no option to change that as far as I know (I think that's issue #1379).

Comment (skyline75489) [reply src/cascadia/TerminalCore/TerminalApi.cpp:393]: Huh. So you can turn off the blinking using VT escape codes. But you can not do that using plain settings. That's just...bizarre,

Update: I almost forgot that I had a PR #2892 which was trying to fix this. How time flies.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (DHowett) [inline src/terminal/adapter/adaptDispatch.cpp:2163]: if i recall conhost correctly, this won't do what we're intending. This will map `0` to legacy (?) even though there's an experimental setting (property sheet, "terminal" page) to set the default cursor style.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (j4james) [inline src/terminal/adapter/adaptDispatch.cpp:2166]: I don't think the behaviour for conhost is right. We should either leave `UserDefault` as an alias for `BlinkingBlock`, which is at least compatible with the DEC standard, or we should try and map it to the actual user preference (which I'm honestly not sure how to do). Worst case we could possibly map it to blinking legacy, which I think is the system default. But as it stands, it looks like you're going to get non-blinking legacy, which is neither one thing nor the other.

Comment (DHowett) [reply src/terminal/adapter/adaptDispatch.cpp:2166]: More detail = more good

**Comment by skyline75489** [comment]
This may be irrelavant but when I'm using stock `cmd.exe` and then `wsl.exe`, the cursor will restore to default after vim exits. I don't fully get it why..

File: src/terminal/adapter/adaptDispatch.cpp
Comment (skyline75489) [reply src/terminal/adapter/adaptDispatch.cpp:2163]: Does `experimental` means that it's only available after certain windows build?

**Comment by DHowett** [comment]
>the cursor will restore to default after vim exits

vim was written for versions of windows that did not support VT, and therefore it almost exclusively uses the _traditional console APIs_. This includes a custom reimplementation of alternate buffers and manual cursor management.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (DHowett) [reply src/terminal/adapter/adaptDispatch.cpp:2163]: It's available in 18362 and higher. https://devblogs.microsoft.com/commandline/new-experimental-console-features/

**Comment by skyline75489** [comment]
@DHowett  But I'm using vim in WSL, right? So it's a Linux Vim, not win32 vim.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (skyline75489) [reply src/terminal/adapter/adaptDispatch.cpp:2163]: I found the settings struct called `CONSOLE_STATE_INFO` but it's not exposed to `congetset`, isn't it.

See this is what happens when you found a "Starter" issue turns out to be a "Pro Plus" issue.

I'm kidding. Give me those "Enterprise" issues next time.

**Comment by DHowett** [comment]
> using WSL vim

oh, right. whoops. :)

I do not believe that it restores the cursor state by default.

**Comment by zadjii-msft** [comment]
I'm gonna hold off on merging this for a bit to give @j4james a chance to voice any objections, since he's still on "requested changes".

File: src/cascadia/TerminalCore/TerminalApi.cpp
Comment (j4james) [inline src/cascadia/TerminalCore/TerminalApi.cpp:421]: Nit: If we're now treating unrecognised parameters as a no-op in Windows Terminal (which is good), should we not be doing the same thing for conhost in AdaptDispatch?

**Review by j4james — APPROVED** [review]
Sorry for the delay. I only have one nit, and it's not a big deal, so I'm happy to approve this.

File: src/cascadia/TerminalCore/TerminalApi.cpp
Comment (skyline75489) [reply src/cascadia/TerminalCore/TerminalApi.cpp:421]: I don’t know how can we debug conhost now. So I intended to keep this. By the way do we want to fix this kind of small bug in conhost, or do we want to keep its behavior. I mean conhost itself is the legacy thing that shouldn’t be changed much for compatibility

File: src/host/outputStream.cpp
Comment (skyline75489) [inline src/host/outputStream.cpp:525]: If we’re on older Windows 10 or Windows 7, will this still work? I assume it will always return legacy.

File: src/cascadia/TerminalCore/TerminalApi.cpp
Comment (j4james) [reply src/cascadia/TerminalCore/TerminalApi.cpp:421]: None of the VT code in conhost is legacy, and this area of the code base is changing all the time. But we can always fix it in a follow-up PR if you don't want to do it now.

Comment (skyline75489) [reply src/cascadia/TerminalCore/TerminalApi.cpp:421]: Thanks. I’m a little uncomfortable that I cannot unittest or manually test conhost. So I don’t feel right about changing that.

I think there’s no need to rush this. And Dustin is not available the next several days. I’m OK waiting for more voices and comments

Comment (j4james) [reply src/cascadia/TerminalCore/TerminalApi.cpp:421]: If you look in the _Conhost_ folder of the of the OpenConsole solution, there's a _Host.EXE_ project that you can run to debug conhost. In the _Debugging_ section of the properties, the _Command Arguments_ specify what shell to use - cmd.exe, bash.exe, etc.

As for the conhost unit testing, it'll depend on what exactly you're doing, but often that will be handled in [ScreenBufferTests.cpp](/microsoft/terminal/blob/master/src/host/ut_host/ScreenBufferTests.cpp) and/or [adapterTest.cpp](/microsoft/terminal/blob/master/src/terminal/adapter/ut_adapter/adapterTest.cpp).

File: src/terminal/adapter/DispatchTypes.hpp
Comment (j4james) [reply src/terminal/adapter/DispatchTypes.hpp]: FYI, when you've resolved an issue that I've raised, and you want to tidy up the comments, you can mark the thread as resolved yourself if you want to. I don't actually have the ability to resolve issues on someone else's PR - I think that's something only MS devs can do.

Comment (skyline75489) [reply src/terminal/adapter/DispatchTypes.hpp]: Yes. Thanks @j4james. I keep forgetting this "resolving comments" feature. I think I got the ability to do this after Dustin granted me the authority, not because I'm a MS dev. BTW I work for Azure in MS, not Console team. So strictly speaking I'm still a community member 😉 .

File: src/terminal/adapter/adaptDispatch.cpp
Comment (skyline75489) [inline src/terminal/adapter/adaptDispatch.cpp:2196]: I think it's fair to treat parameter larger than 6 as a "success". I mean, why bother sending it to the connected terminal when we know it should be ignored.

Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:2196]: Well now that you mention it, technically we should probably be passing through any unknown parameters. The connected terminal is not necessarily WT, and it could potentially support values larger than 6. We were actually just discussing this in issue https://github.com/microsoft/terminal/issues/7382#issuecomment-683318287. I don't think it's a big deal, but returning false here is probably better (assuming that doesn't cause any other problems).

File: src/host/outputStream.cpp
Comment (j4james) [reply src/host/outputStream.cpp:525]: I believe the VT support was only added to conhost in build 10586 of Windows 10, so nothing prior to that should have any of this code. Maybe one of the core devs can confirm this, but I don't think we need to worry about older versions of Windows at all.

File: src/terminal/adapter/adaptDispatch.cpp
Comment (skyline75489) [reply src/terminal/adapter/adaptDispatch.cpp]: That makes sense. I wasn't really sure about it. Let's try 'false' for this.

File: src/host/outputStream.cpp
Comment (DHowett) [reply src/host/outputStream.cpp:525]: Yeah, this is totally fine. Conhost is also shipped as a totally independent unit -- so if conhost can run on an older version of Windows, it will bring with it **full VT support** and this cursor change.

(incidentally, it works on Windows 8.1 and brings full VT support :wink:)

File: src/terminal/adapter/ut_adapter/adapterTest.cpp
Comment (DHowett) [inline src/terminal/adapter/ut_adapter/adapterTest.cpp:440]:
```suggestion
        return true;
```

**Review by DHowett — APPROVED** [review]
This is excellent. Thanks!

**Comment by younger-1** [comment]
In windows terminal, this issues is still 

> vim does not attempt to restore cursor shape on exit

vim and neovim both do not restore cursor on exit if without configuring vimrc...

**Comment by ubaldot** [comment]
> > I don't think this actually _fixes_ #1604 as vim does not attempt to restore cursor shape on exit. So..what do you guys think?
> 
> neovim and vim both do not restore cursor on exit if without configuring vimrc...

I bumped into the same issue: shall I then explicitly use `au VimLeave * set guicursor=a:hor10-blinkon1`in my vimrc to get my cursor back? 

Also, I have noticed that when accessing Vim the cursor won't take the color of the used colorscheme. Any idea why and how to fix it?

**Pull request merged**
